### ISSUE TITLE
Turning TCK into a multi-dependency maven project

### ISSUE THREAD

**Issue by mkarg** [jakartaee/platform-tck#51]
To make contributing and using the TCK easier for those working only on a small subset of APIs without making it overly hard for those who work on Jakarta EE as a full set of APIs, I'd like to propose to turn the TCK project into a Maven multi-dependency assembly project:
* Split up the monolith into TCK-per-API and move the TCKs into the repositories of the particular API projects.
* The API projects compile and release the TCKs together with their APIs to ensure that API and TCK always fit together, so the TCKs are found on Maven Central at the same time the APIs appear there. This will be used by single-API-vendors (like EclipseLink, Jersey, etc.).
* The original (ex-monolith) TCK project provides a POM which pulls all those single TCKs from Maven Central as dependencies and packs them together in a single-shot umbrella assembly. This will be used by profile vendors (like GlassFish, Payara, etc.).

**Comment by bshannon** [comment]
Looking forward to your pull request!  :-)

In general, something like this is the right approach.  Note that there's a bunch of common code that
will need to be refactored into a shareable repository as well.

**Comment by scottmarlow** [comment]
Do you mean all of the tests or only the standalone TCK tests?

If you mean all of the tests, I'd like to see discussion on jakartaee-tck-dev@eclipse.org about this, regarding the general approach to take for moving the TCK tests, especially around deciding what to do with the many tests that span multiple technologies.  

Agreed about the common code moving into a shareable repository.

**Comment by mkarg** [comment]
@bshannon Would be happy to provide a PR, but I will definitively need lots of help as I have not used the TCK ever before.

@scottmarlow My plan is to selectively extract particular tests API-by-API with the help of the particular API teams, i. e. like JAX-RS TCK, JMS TCK, etc. step by step. There might be tests left over. Certainly there has to happen discussions with all committers. Why do you want this discussion outside of this issue? It makes harder to participate for Github-but-not-Eclipse people.

**Comment by bshannon** [comment]
Anyone can subscribe to the mailing list, why is that a problem?

**Comment by scottmarlow** [comment]
@mkarg it will be difficult to contain the entire discussion here in this one issue tracker.  I think that we need to discuss this with a threaded forum or mailing list, not a (flat github issue).

**Comment by mkarg** [comment]
@scottmarlow I do not share any of that thoughts, but on the other hand, it is not me to decide. Once you have a final decision I would be happy if you write down the outcome here. :-)

**Comment by scottmarlow** [comment]
@mkarg Thanks for the opportunity. :-)  

Before we start the mailing list discussion, could you give details of what the likely impact will be on the current sources like [https://github.com/eclipse-ee4j/jakartaee-tck/tree/master/src/com/sun/ts/tests/jaxrs], as that contains tests for the Jakarta JAX-RS standalone TCK as well as Jakarta EE 8 TCK.  Is the idea that tests that require a Jakarta EE 8 server implementation to run, would likely be the tests that are left over in the jakartaee-tck repo?

**Comment by scottmarlow** [comment]
Hi @mkarg a lot has changed since April of 2019.  We have at least 6 conversations going on now about different ways to create  SPEC API level TCKs for certain SPEC API projects (mostly the SPEC API projects that are looking at making SPEC changes).  This particular issue for updating the Platform TCK into a mult-dependency maven project is something that I would like to start working on.  As well as to facilitate the SPEC API projects interested in developing their next SPEC API version using their own TCK.  I started working on collecting the list of dependencies for each Platform TCK test group (including the number of Platform level tests).  

The work in progress script for collecting the TCK info is at https://gist.github.com/scottmarlow/d4a13d15e7c0ea83d35004445b4b84f0 and a now out of date (from last week) example output is at https://gist.github.com/scottmarlow/0e31e5256507ad0e5c78b3f274d8d4a7.  I started working a global change to remove unused Java imports to eliminate false dependencies from the report.  Removing unused imports is going well via https://github.com/eclipse-ee4j/jakartaee-tck/pull/582 but I need to update the copyright dates still for that pr still.

I will start an email conversation soon with all of the SPEC API mailing lists and encourage people to respond all.  This will mean that we have a number of separate per SPEC API conversations.  I'm open to suggestions for which other mailing lists that we could include that all might be members of.
